improver/utilities/temporal_interpolation.py

```diff
@@ -684,3 +684,242 @@ def process(self, cube_t0: Cube, cube_t1: Cube) -> CubeList:
+            day_mask:
+                If true, points that fall in the day time are zeroed and
+                duration reallocated to night time periods as much as possible.
+        Raises:
+            ValurError: If day and night mask options are both set True.
+        """
+        for item in [target_period, fidelity]:
+            if item <= 0:
+                raise ValueError(
+                    "Target period and fidelity must be a positive integer "
+                    "numbers of seconds. Currently set to "
+                    f"target_period: {target_period}, fidelity: {fidelity}"
+                )
+
+        self.target_period = target_period
+        self.fidelity = fidelity
+        if night_mask and day_mask:
+            raise ValueError(
+                "Only one or neither of night_mask and day_mask may be set to True"
+            )
+        elif not night_mask and not day_mask:
+            self.mask_value = None
+        else:
+            self.mask_value = 0 if night_mask else 1
+
+    @staticmethod
+    def cube_period(cube):
+        """Return the time period of the cube in seconds.
+
+        Args:
+            cube:
+                The cube for which the period is to be returned.
+        Return:
+            period:
+                Period of cube time coordinate in seconds.
+        """
+        (period,) = np.diff(cube.coord("time").bounds[0])
+        return period
+
+    def allocate_data(self, cube, period):
+        """Allocate fractions of the total duration to the shorter periods
+        and modify the metadata of the shorter period cubes appropriately.
+
+        Args:
+            cube:
+                The original period cube from which duration data will be
+                taken and divided up.
+            period:
+                The target shorter period
+        Returns:
+            A cube, with a time dimension, that contains the subdivided data.
+        """
+        # Split the whole period duration into allocations for each fidelity
+        # period.
+        intervals = period // self.fidelity
+        interval_data = cube.data / intervals
+
+        daynightplugin = DayNightMask()
+        start_time, _ = cube.coord("time").bounds.flatten()
+
+        interpolated_cubes = iris.cube.CubeList()
+
+        for i in range(intervals):
+            interval_cube = cube.copy(data=interval_data.copy())
+            interval_start = start_time + i * self.fidelity
+            interval_end = start_time + (i + 1) * self.fidelity
+
+            interval_cube.coord("time").points = np.array(
+                [interval_end], dtype=np.int64
+            )
+            interval_cube.coord("time").bounds = np.array(
+                [[interval_start, interval_end]], dtype=np.int64
+            )
+            daynight_mask = daynightplugin(interval_cube).data
+            daynight_mask = np.broadcast_to(daynight_mask, interval_cube.shape)
```

Comment:
Perhaps these lines should be moved into the if block below, as the daynightplugin only needs to be called if the mask needs to be applied? E.g.

```
if self.mask_value is not None:
    daynight_mask = daynightplugin(interval_cube).data
    daynight_mask = np.broadcast_to(daynight_mask,  interval_cube.shape)
    interval_cube.data[daynight_mask == self.mask_value] = 0.0
```